tests/testthat/test-compare.R

```diff
@@ -107,6 +147,59 @@ test_that("vec_proxy_compare() handles data frame with a POSIXlt column", {
   )
 })
 
+test_that("error is thrown with data frames with 0 columns", {
+  x <- new_data_frame(n = 1L)
+  expect_error(vec_compare(x, x), "data frame with zero columns")
+})
+
+test_that("error is thrown when comparing lists", {
+  expect_error(vec_compare(list(), list()), class = "vctrs_error_unsupported")
+  expect_error(.Call(vctrs_compare, list(), list(), FALSE), "Can't compare lists")
+})
+
+test_that("error is thrown when comparing scalars", {
+  x <- new_sclr(x = 1)
+  expect_error(vec_compare(x, x), class = "vctrs_error_scalar_type")
+  expect_error(.Call(vctrs_compare, x, x, FALSE), class = "vctrs_error_scalar_type")
+})
+
+test_that("`na_equal` is validated", {
+  expect_error(vec_compare(1, 1, na_equal = 1), class = "vctrs_error_assert_ptype")
+  expect_error(vec_compare(1, 1, na_equal = c(TRUE, FALSE)), class = "vctrs_error_assert_size")
+})
+
+test_that("can compare equal strings with different encodings", {
+  for (x_encoding in encodings()) {
+    for (y_encoding in encodings()) {
+      expect_equal(vec_compare(x_encoding, y_encoding), 0L)
+    }
+  }
+})
+
+test_that("can compare different strings with different encodings", {
+  x <- "x"
+  y <- encodings()$latin1
+
+  expect_equal(vec_compare(x, y), -1L)
+})
+
+test_that("equality can always be determined when strings have identical encodings", {
+  encs <- c(encodings(), list(bytes = encoding_bytes()))
```

Comment:
I was puzzled for a bit here. I wonder if we should switch to this clearer way of combining lists:

```
encs <- list2(!!!encodings(), bytes = encoding_bytes())
```